### ISSUE TITLE
Specify descendant or ancestor in query

### ISSUE THREAD

**Issue by robbym** [tree-sitter/tree-sitter#880]
Is there a way to pattern match a descendant of arbitrary depth?

I'm parsing C code and I'm trying to get all identifiers of static declarations.
Something like the following works for simple cases:

Given:
```c
static int test;
static int test = 1;
static int test();
static int* test;
```
This works:
```
(declaration
    (storage_class_specifier "static")
    declarator: [
        (identifier)@name
        (_ declarator: (identifier)@name)
    ]
)
```

However, with more complex declarations:
```c
static int* test();
static int** test;
static int*** test;
```
Require me to keep adding alternates:
```
(declaration
    (storage_class_specifier "static")
    declarator: [
        (identifier)@name
        (_ declarator: (identifier)@name)
        (_ declarator: (_ declarator: (identifier)@name))
        (_ declarator: (_ declarator: (_ declarator: (identifier)@name))
        ...
    ]
)
```

Is there any way to specify the identifier whose ancestor (of arbitrary depth) is a static declaration (or declaration in general)?

**Comment by robbym** [comment]
Maybe something like introducing a subquery syntax?

```
(declaration (storage_class_specifier "static") => (_ declarator: (identifier)@name)
```

This basically reads as: For each declaration with a static storage_class_specifier, match any "sub-declarator" that is an identifier.

**Comment by robbym** [comment]
Or maybe something like:
```
(declaration
    (storage_class_specifier "static")
    { (_ declarator: (identifier)@name) }
)
```

Where anything in a set of braces is just an arbitrary query, with the only exception that the root node being traversed by said subquery must match the parent pattern.

**Comment by MunifTanjim** [comment]
> I think all mentioned cases with catastrophic recursion actually have an issue in a grammars' composition and instead of attempts to handle unlimited recursions by queries it's better to reorganize grammars and flatten recursions by replacing a recursive rule by a rule with a repetition of the part that produces recursion.

Most of the parsers for popular languages in under https://github.com/tree-sitter org is made that way. That makes it hard to argue that it's not the recommended way of doing it.

**Comment by tmillr** [comment]
> > I think all mentioned cases with catastrophic recursion actually have an issue in a grammars' composition and instead of attempts to handle unlimited recursions by queries it's better to reorganize grammars and flatten recursions by replacing a recursive rule by a rule with a repetition of the part that produces recursion.
> 
> Most of the parsers for popular languages under https://github.com/tree-sitter org is made that way. That makes it hard to argue that it's not the recommended way of doing it.

True. However, when it comes to index expressions, I feel like it could go either way perhaps? On the other hand, if something like the highly nested HTML/XML (e.g. elements) were converted into sequences of sibling nodes somehow, that would feel very wrong (and probably because just about everything in HTML is (highly) nested). So in that case, supporting the feature discussed herein would probably require the treesitter query grammar/language to be extended anyway in order to support these sort of recursive patterns/expressions. And it's probably more correct/semantically correct to keep index expressions as they are currently. For example, Lua appears to specify/model them as recursive (which is in-line with how the code executes):

```
var       ::= Name | prefixexp ‘[’ exp ‘]’ | prefixexp ‘.’ Name
prefixexp ::= var  | functioncall          | ‘(’ exp ‘)’
```

To convert index expressions into sequences, I imagine that it'd also require a new/additional container node in order to hold all of the siblings under a single field (e.g. note the fields for `dot_index_expression` from my previous post) (or simply ditching this field altogether), and you'd probably lose the ability to reference each of the siblings separately via a field.

**Comment by the-xentropy** [comment]
If you need to find N nodes nested in each other, you can do it with N queries and some magic dust. Let me show you two generic ways to do this.

Problem example:

```
int foo() {
   // printf() displays the string inside quotation
   foobar(foofoobar("Hi!"));
   return 0;
}
```
I wish to find all declared functions, and all functions called within it. So I simultaneously want the information 'foo' as well as 'foobar' and 'foofoobar'.

First query to find function declarations ('find_function_declarations_query'):
```
(function_definition
    declarator: (function_declarator
        declarator: (identifier) @function.name
    ))  @function.node
```

Second query to find calls ('find_function_calls_query'):
```
(call_expression
    function: (identifier) @call.name)
```

If we combine them into a single query, as the bug correctly points out we only get one layer. 

However, if we use the fact that .query() returns the full node on each match, we can solve our problem:

```python
first_layer_matches = find_function_declarations_query.captures(tree.root_node)
second_layer_matches = [item for sublist in [find_function_calls_query.captures(match[0]) for match in  first_layer_matches if match[1] == 'function.node'] for item in sublist]
print(second_layer_matches)
print([match[0].text for match in second_layer_matches if match[1] == 'call.name'])
```

The result is `[b'foobar', b'foofoobar']` . Which is what we wanted.

Making a generic version of this:

```python
def subquery(root_node: Node, queries: Tuple[Query, ...]) -> Tuple[Tuple, ...]:
    """
    query with support for arbitrary depth. Only returns terminating nodes.

    root_node: Node, root node of the search
    queries: Tuple[Query, ...], queries that need to be executed at any depth but always as sub-queries of the previous executed query
    """
    if len(queries) == 0:
        return () # base case, no result for no query
    last_results = None
    for query in queries:
        if last_results is None:
            last_results = query.captures(root_node)
            continue # we have our 1st layer results, we can continue with the next query
        subquery_results = []
        for old_result in last_results:
            node, _ = old_result
            subquery_result = query.captures(node)
            if len(subquery_result) == 0:
                continue # no result for this subquery, no need to continue
            subquery_results.extend(subquery_result)
        last_results = subquery_results
    return last_results
print(subquery(tree.root_node, (find_function_declarations_query, find_function_calls_query)))
# [b'foobar', b'foofoobar']
```


If you want to save information at different steps (e.g. name of the function declaration and the calls in the function), you'll have to be a bit more clever because the Node object doesn't have a \_\_hash\_\_ implementation and won't work as keys in dicts/sets and so we can't just make an arbitrary subtree with just findings.

As a easy generic for what I assume to be about 90% of this use-case (one parent, many potential places for termination) where you might want info from both the parent and final descendant:

```python
def subquery_with_parent_info(root_node: Node, queries: Tuple[Query, ...]) -> List[Tuple]:
    """
    query with support for arbitrary depth

    root_node: Node, root node of the search
    queries: Tuple[Query, ...], queries that need to be executed at any depth but always as sub-queries of the previous executed query
    """
    if len(queries) == 0:
        return () # base case, no result for no query
    final_result = []
    parent_query = queries[0]
    parent_results = parent_query.captures(root_node)
    for parent_result in parent_results:
        parent_node, _ = parent_result
        res = subquery(parent_node,queries[1:])
        if len(res) > 0:
            final_result.append((parent_result, subquery(parent_node,queries[1:])))
    return final_result
```
which returns something like `[((<Node type=function_definition, start_point=(0, 0), end_point=(4, 1)>, 'function.node'), [(<Node type=identifier, start_point=(2, 3), end_point=(2, 9)>, 'call.name'), (<Node type=identifier, start_point=(2, 10), end_point=(2, 19)>, 'call.name')])] `

Keeping this functionality out of tree sitter itself is probably a good idea though. You end up having to create various variations of the same solutions depending on your output format needs. Potentially what could work is returning a tree representing the tree of partial answers (e.g. AnswerNodes with other AnswerNodes as its children) but... I think most people want to avoid tree structures, so it creates another problem.

Anyway, just documenting both for my own future reference as well as for anyone else stumbling into this issue in the future. :)

**Comment by omentic** [comment]
I'm a bit surprised there isn't more interest in this! I've run into this issue with nearly every (every?) grammar I've written queries for, specifically with nested types in generics or chained function types - the only options are to (incorrectly) highlight the entire type expression, or only highlight up to N number of nested types, neither of which are very good solutions. 

If there isn't horrible performance degradation this'd be very helpful.

**Comment by sangwoo-joh** [comment]
Any update with this issue?

I'm not sure if there's a way to bypass this by positively capturing the `*` as anonymous nodes, but I don't get the right way yet.

**Comment by vergenzt** [comment]
Could introducing a "recursive descent" operator work? E.g. a new `..` symbol, indicating that the remaining child pattern(s) in the list should be checked for a match against the remaining child node(s) _or any of their descendants_, instead of having to match a subsequence of the remaining direct child nodes.

I like `..` because
1. it's stylistically comparable to [anchors which use a single period (`.`)](https://tree-sitter.github.io/tree-sitter/using-parsers/queries/2-operators.html#anchors); and
2. it's reminiscent to me of [jq's recursive descent operator `..`](https://jqlang.org/manual/#recursive-descent)).

Obviously we could choose a different symbol though.

E.g. something like
```
(declaration
    (storage_class_specifier "static")
    ..
    declarator: (identifier) @name
)
```

**Comment by greneholt** [comment]
Another case where this feature would be useful is in [SystemVerilog](https://github.com/gmlarumbe/tree-sitter-systemverilog/), where I want to color all parameters/constants with a consistent color, including when they are part of constant expressions. But the language grammar re-uses the non-constant function call syntax within constant expressions, so it is impossible to properly select all parameters within constant function calls and color them appropriately.
